python/ray/data/dataset.py

```diff
@@ -613,6 +613,57 @@ def do_shuffle(block_list, clear_input_blocks: bool, block_udf, remote_args):
         )
         return Dataset(plan, self._epoch, self._lazy)
 
+    def random_sample(self, number: int, *, seed: Optional[int] = None) -> List[Any]:
+        """Randomly samples N elements from the dataset.
+
+        Examples:
+            >>> import ray
+            >>> ds = ray.data.range(100) # doctest: +SKIP
+            >>> ds.random_sample(5) # doctest: +SKIP
+            >>> # Sample this dataset with a fixed random seed.
+            >>> ds.random_sample(5, seed=12345) # doctest: +SKIP
+
+
+        Args:
+            number: The number of elements to sample from the dataset.
+
+            seed: Seeds the python random pRNG generator.
+
+        Returns:
+            N elements, randomly sampled from the dataset.
+        """
+        import random
+
+        if self.num_blocks() == 0:
+            raise ValueError("Cannot from an empty dataset")
+
+        if number < 1:
+            raise ValueError("Cannot sample less than 1 element.")
+
+        count = self._meta_count()
+
+        if number > count:
+            raise ValueError(
+                "Cannot sample more elements than there are in the dataset"
+            )
+
+        if seed:
+            random.seed(seed)
+
+        n_required = number // self.num_blocks()
```

Comment:
That's why I added a check below. For a dataset with _r_ rows, I noticed the implementation split it into _n_ blocks with _p_ elements each and 1 block with _q_ elements where _q<p_, and _p*n+q=r_. I simply added 1 if there was a remainder so this will actually pull more elements than "needed", and truncates at _N_ elements.

Reply:
For example, this code works fine:
```python
import ray

ds = ray.data.range(100, parallelism=8)
# splits into 8 blocks, 7 blocks with 14 and 1 block with 2
sample = ds.random_sample(17)
# 17 // 8 = 2
# since 17%8 !=0 we just add 1 
# so sample min(num_rows_in_that_block, 2+1)
# since the last block only has 2 it will only sample 2
assert len(sample) == 17
```

Reply:
So the issue is that the uneven distribution of rows may not be just due to the indivisibility. For example, it may have 4 blocks with num rows 20, 5, 20, 5.

Reply:
Okay, I'll work on a better sampling algorithm.

Reply:
One potential algorithm is reservoir sampling (https://en.wikipedia.org/wiki/Reservoir_sampling) which can just linear scan each block (and doesn't need to know how many rows in the block or in dataset).

Reply:
Hmm, this could be tricky since you don't want to introduce bias towards or against smaller blocks. Maybe the original idea of generating the sample indices up front, and then pulling them from the blocks at the appropriate index, is the best way here.

To generate such blocks you could .union() datasets of different sized blocks.

Reply:
Ok, actually Pandas uses this approach (sample indices and then take them out): https://github.com/pandas-dev/pandas/blob/v1.4.2/pandas/core/generic.py#L5299-L5452

@xiurobert One way I think should be able to generate such blocks is to create 4 files (can be text/csv if you want, with the different number of rows) and then load them into a Dataset.

Reply:
Yup I'm writing some tests now.